AX: AXCoreObject::{nextSiblingIncludingIgnored, previousSiblingIncludingIgnored} are O(n), causing poor performance on sites with big accessibility trees

twilco opened the pull request:
#### 202b4702d12b320ab03a6c2b9e4be7918023ab78
<pre>
AX: AXCoreObject::{nextSiblingIncludingIgnored, previousSiblingIncludingIgnored} are O(n), causing poor performance on sites with big accessibility trees
<a href="https://bugs.webkit.org/show_bug.cgi?id=292799">https://bugs.webkit.org/show_bug.cgi?id=292799</a>
<a href="https://rdar.apple.com/151029465">rdar://151029465</a>

Reviewed by Joshua Hoffman.

Prior to this commit, AXCoreObject::{nextSiblingIncludingIgnored, previousSiblingIncludingIgnored} were O(n) because
it iterated over siblings to find |this|, then used that index to return the next or previous sibling.

With this commit, each object now has an m_indexInParent member variable that is assigned by its parent when added as
a child. This new member variable does not increase the size of AXCoreObject, AXIsolatedObject, or AccessibilityObject
at all, as it was placed into a place that used to be occupied by padding between existing member variables.

This commit adds ASSERTs that verifies the cached indices are correct (contiguous, starting from zero, no gaps). When
debugging one of these ASSERTs, I discovered another performance optimization that also prevents said ASSERT. Prior to
this commit, we used to add all DOM children, then add our list marker child at index 0. This requires shifting the
entire vector over, which is O(n). Now, we move the addListItemMarker() call to the right place, obviating the need
for this O(n) work.

This eliminates 23782 samples of 49643 accessibility thread samples on <a href="http://html.spec.whatwg.org">http://html.spec.whatwg.org</a>, taken while holding
the VO-Right keys.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::verifyChildrenIndexInParent const):
(WebCore::AXCoreObject::indexInSiblings const):
(WebCore::AXCoreObject::nextSiblingIncludingIgnored const):
(WebCore::AXCoreObject::previousSiblingIncludingIgnored):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::setIndexInParent):
(WebCore::AXCoreObject::shouldSetChildIndexInParent const):
(WebCore::AXCoreObject::setChildIndexInParent const):
(WebCore::AXCoreObject::indexInParent const):
* Source/WebCore/accessibility/AccessibilityListBox.cpp:
(WebCore::AccessibilityListBox::addChildren):
* Source/WebCore/accessibility/AccessibilityMathMLElement.cpp:
(WebCore::AccessibilityMathMLElement::addChildren):
* Source/WebCore/accessibility/AccessibilityMenuListPopup.cpp:
(WebCore::AccessibilityMenuListPopup::addChildren):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::updateOwnedChildren):
(WebCore::AccessibilityNodeObject::addChildren):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::insertChild):
(WebCore::AccessibilityObject::resetChildrenIndexInParent const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::addChildren):
(WebCore::AccessibilityRenderObject::addListItemMarker): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::removeChildScrollbar):
(WebCore::AccessibilityScrollView::addChildren):
* Source/WebCore/accessibility/AccessibilitySlider.cpp:
(WebCore::AccessibilitySlider::addChildren):
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::addChildren):
* Source/WebCore/accessibility/AccessibilityTableColumn.cpp:
(WebCore::AccessibilityTableColumn::addChildren):
* Source/WebCore/accessibility/AccessibilityTableHeaderContainer.cpp:
(WebCore::AccessibilityTableHeaderContainer::addChildren):
* Source/WebCore/accessibility/AccessibilityTableRow.cpp:
(WebCore::AccessibilityTableRow::addChildren):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::children):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:

Canonical link: <a href="https://commits.webkit.org/294747@main">https://commits.webkit.org/294747@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3fbfc8dcad204d30a9750dd7e2e930a8e1560c92

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102888 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22562 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12882 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108054 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53529 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104927 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22882 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31063 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78223 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35178 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105894 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17719 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92850 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58557 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17559 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10885 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52886 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87368 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10952 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110429 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30025 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22118 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87209 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30389 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89045 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86828 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22111 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31678 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9393 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24287 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29952 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35274 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29760 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33087 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31322 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -687,21 +692,33 @@ void AccessibilityObject::insertChild(AccessibilityObject& child, unsigned index
+
+    unsigned index = 0;
+    for (const auto& child : m_children) {
+        if (setChildIndexInParent(child.get(), index))
```
Since you're already checking shouldSetChildIndexInParent, this should never return false, right? Could you make it a debug check instead?

twilco replied:
Ahh, great point! Yeah I'll turn it into an ASSERT.

minorninth commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -671,6 +671,11 @@ void AccessibilityObject::insertChild(AccessibilityObject& child, unsigned index
 #endif // USE(ATSPI)
 
+    auto insert = [this] (Ref<AXCoreObject>&& object, unsigned index) {
+        setChildIndexInParent(object.get(), index);
```
Is this one ignoring the return value? As mentioned in the other comment, maybe it should just be a debug check? Or alternatively maybe use std::ignore to make it explicit that you're ignoring the return value?

twilco replied:
Yeah, in this case we intentionally ignore the return value, since the insertion index is managed elsewhere in this function. But `std::ignore` would be good to make it explicit, I'll do that.

minorninth reviewed (approved):
Great optimization!

twilco committed:
AX: AXCoreObject::{nextSiblingIncludingIgnored, previousSiblingIncludingIgnored} are O(n), causing poor performance on sites with big accessibility trees
https://bugs.webkit.org/show_bug.cgi?id=292799
rdar://151029465

Reviewed by Joshua Hoffman.

Prior to this commit, AXCoreObject::{nextSiblingIncludingIgnored, previousSiblingIncludingIgnored} were O(n) because
it iterated over siblings to find |this|, then used that index to return the next or previous sibling.

With this commit, each object now has an m_indexInParent member variable that is assigned by its parent when added as
a child. This new member variable does not increase the size of AXCoreObject, AXIsolatedObject, or AccessibilityObject
at all, as it was placed into a place that used to be occupied by padding between existing member variables.

This commit adds ASSERTs that verifies the cached indices are correct (contiguous, starting from zero, no gaps). When
debugging one of these ASSERTs, I discovered another performance optimization that also prevents said ASSERT. Prior to
this commit, we used to add all DOM children, then add our list marker child at index 0. This requires shifting the
entire vector over, which is O(n). Now, we move the addListItemMarker() call to the right place, obviating the need
for this O(n) work.

This eliminates 23782 samples of 49643 accessibility thread samples on http://html.spec.whatwg.org, taken while holding
the VO-Right keys.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::verifyChildrenIndexInParent const):
(WebCore::AXCoreObject::indexInSiblings const):
(WebCore::AXCoreObject::nextSiblingIncludingIgnored const):
(WebCore::AXCoreObject::previousSiblingIncludingIgnored):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::setIndexInParent):
(WebCore::AXCoreObject::shouldSetChildIndexInParent const):
(WebCore::AXCoreObject::setChildIndexInParent const):
(WebCore::AXCoreObject::indexInParent const):
* Source/WebCore/accessibility/AccessibilityListBox.cpp:
(WebCore::AccessibilityListBox::addChildren):
* Source/WebCore/accessibility/AccessibilityMathMLElement.cpp:
(WebCore::AccessibilityMathMLElement::addChildren):
* Source/WebCore/accessibility/AccessibilityMenuListPopup.cpp:
(WebCore::AccessibilityMenuListPopup::addChildren):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::updateOwnedChildren):
(WebCore::AccessibilityNodeObject::addChildren):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::insertChild):
(WebCore::AccessibilityObject::resetChildrenIndexInParent const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::addChildren):
(WebCore::AccessibilityRenderObject::addListItemMarker): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::removeChildScrollbar):
(WebCore::AccessibilityScrollView::addChildren):
* Source/WebCore/accessibility/AccessibilitySlider.cpp:
(WebCore::AccessibilitySlider::addChildren):
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::addChildren):
* Source/WebCore/accessibility/AccessibilityTableColumn.cpp:
(WebCore::AccessibilityTableColumn::addChildren):
* Source/WebCore/accessibility/AccessibilityTableHeaderContainer.cpp:
(WebCore::AccessibilityTableHeaderContainer::addChildren):
* Source/WebCore/accessibility/AccessibilityTableRow.cpp:
(WebCore::AccessibilityTableRow::addChildren):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::children):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:

Canonical link: https://commits.webkit.org/294747@main
